components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/PreferenceRetrievalClient.java

```diff
@@ -0,0 +1,263 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.wso2.carbon.identity.mgt.endpoint.util.client;
+
+import org.apache.axis2.transport.http.HTTPConstants;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.log4j.Logger;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+import org.wso2.carbon.identity.mgt.endpoint.util.IdentityManagementEndpointUtil;
+import org.wso2.carbon.identity.mgt.endpoint.util.IdentityManagementServiceUtil;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Client which retrieves preferences.
+ */
+public class PreferenceRetrievalClient {
+
+    private static final String CLIENT = "Client ";
+    private static final String PROPERTIES = "properties";
+    private static final Logger log = Logger.getLogger(PreferenceRetrievalClient.class);
+    private static final String PREFERENCE_API_RELATIVE_PATH = "/api/server/v1/identity-governance/preferences";
+    private static final String SELF_REGISTRATION_PROPERTY = "SelfRegistration.Enable";
+    private static final String USERNAME_RECOVERY_PROPERTY = "Recovery.Notification.Username.Enable";
+    private static final String NOTIFICATION_PASSWORD_RECOVERY_PROPERTY = "Recovery.Notification.Password.Enable";
+    private static final String QUESTION_PASSWORD_RECOVERY_PROPERTY = "Recovery.Question.Password.Enable";
+    private static final String CONNECTOR_NAME = "connector-name";
+    private static final String SELF_SIGN_UP_CONNECTOR = "self-sign-up";
+    private static final String RECOVERY_CONNECTOR = "account-recovery";
+    private static final String PROPERTY_NAME = "name";
+    private static final String PROPERTY_VALUE = "value";
+
+    /**
+     * Check self registration is enabled or not.
+     *
+     * @param tenant tenant domain name.
+     * @return returns true if self registration enabled.
+     * @throws PreferenceRetrievalClientException
+     */
+    public Boolean checkSelfRegistration(String tenant) throws PreferenceRetrievalClientException {
```

Comment:
Any particular reason why we opted to return a Boolean object instead of a boolean primitive type?

Reply:
No, we could change it to a boolean primitive type